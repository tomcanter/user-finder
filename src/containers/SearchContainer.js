// @flow

import React, {
  Component,
  PropTypes,
} from 'react';
import SearchForm from '../components/SearchForm';
import connect from './SearchConnect';

type Props = {
  pushRoute: Function,
  searchUser: Function,
  query: string,
};

export class SearchContainer extends Component {

  static propTypes = {
    pushRoute: PropTypes.func.isRequired,
    searchUser: PropTypes.func.isRequired,
    query: PropTypes.string.isRequired,
  };

  constructor(props: Props) {
    super(props);
    this.handleSearchUser(props.query);
  }

  componentWillReceiveProps(nextProps: Props) {
    if (nextProps.query !== this.props.query) {
      this.handleSearchUser(nextProps.query);
    }
  }

  handleSearchUser(query: string): void {
    if (!query) {return;}
    this.props.searchUser({query});
  }

  render() {
    const {
      query,
      pushRoute,
    } = this.props;

    return (
      <div>
        <SearchForm
          pushRoute={pushRoute}
          query={query}
        />
      </div>
    );
  }

}

export default connect(SearchContainer);
