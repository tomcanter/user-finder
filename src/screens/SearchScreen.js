// @flow

import React, {
  Component,
  PropTypes,
} from 'react';
import qs from 'query-string';
import SearchContainer from '../containers/SearchContainer';

type Props = {
  push: Function,
  location: Object,
};

class SearchScreen extends Component {

  props: Props;

  static propTypes = {
    push: PropTypes.func.isRequired,
    location: PropTypes.object.isRequired,
  };

  pushUrlQuery = (value: string): void => {
    this.props.push({
      path: '/',
      search: qs.stringify({
        q: value,
        page: 1,
      }),
    });
  }

  render() {
    const {
      location,
      push,
    } = this.props;
    const parsedSearch = qs.parse(location.search);

    return (
      <div>
        <SearchContainer
          pushRoute={push}
          searchTerm={parsedSearch.q}
          search={qs.stringify(parsedSearch)}
          onSubmit={this.pushUrlQuery}
        />
      </div>
    );
  }

}

export default SearchScreen;
