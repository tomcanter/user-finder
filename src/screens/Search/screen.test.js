import React from 'react';
import {shallow} from 'enzyme';
import SearchScreen from './';
import SearchContainer from '../../containers/Search';

describe('Screen: SearchScreen', () => {

  describe('when passed a location object', () => {
    it('should parse the searchTerm and pass it to SearchContainer', () => {
      const location = {
        search: '?q=testing',
      };
      const wrapper = shallow(
        <SearchScreen
          push={jest.fn()}
          location={location}
        />
      );
      expect(wrapper.find(SearchContainer).first().props().searchTerm).toEqual('testing');
    });
  });

  describe('constructTitle function', () => {
    it('should use the base title when no search query is present', () => {
      const title = SearchScreen.constructTitle({});
      expect(title).toEqual('Github User Search');
    });

    it('should return a title describing current search and page', () => {
      const title = SearchScreen.constructTitle({q: 'foo', page: 2});
      expect(title).toEqual('foo - Page 2 - Github User Search');
    });
  });

  describe('pushUrlQuery function', () => {
    it('should push a search query on the URL', () => {
      const spy = jest.fn();
      const wrapper = shallow(
        <SearchScreen
          push={spy}
          location={{}}
        />
      );
      wrapper.instance().pushUrlQuery('testing');
      expect(spy.mock.calls[0]).toMatchSnapshot();
    });
  });

});
