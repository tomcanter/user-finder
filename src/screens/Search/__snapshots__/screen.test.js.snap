exports[`Screen: SearchScreen pushUrlQuery function should push a search query on the URL 1`] = `
Array [
  Object {
    "path": "/",
    "search": "per_page=45&page=1&q=testing",
  },
]
`;
